cmake/Modules/FindClangFormat.cmake

```diff
@@ -0,0 +1,94 @@
+#
+# .rst: FindClangFormat
+# ---------------
+#
+# The module defines the following variables
+#
+# ``CLANG_FORMAT_EXECUTABLE`` Path to clang-format executable
+# ``CLANG_FORMAT_FOUND`` True if the clang-format executable was found.
+# ``CLANG_FORMAT_VERSION`` The version of clang-format found
+# ``CLANG_FORMAT_VERSION_MAJOR`` The clang-format major version if specified, 0
+# otherwise ``CLANG_FORMAT_VERSION_MINOR`` The clang-format minor version if
+# specified, 0 otherwise ``CLANG_FORMAT_VERSION_PATCH`` The clang-format patch
+# version if specified, 0 otherwise ``CLANG_FORMAT_VERSION_COUNT`` Number of
+# version components reported by clang-format
+#
+# Example usage:
+#
+# .. code-block:: cmake
+#
+# find_package(ClangFormat) if(CLANG_FORMAT_FOUND) message("clang-format
+# executable found: ${CLANG_FORMAT_EXECUTABLE}\n" "version:
+# ${CLANG_FORMAT_VERSION}") endif()
+
+find_program(CLANG_FORMAT_EXECUTABLE
+             NAMES
+                   clang-format-9
+                   clang-format-9.0
+                   clang-format-8
+                   clang-format-8.0
+                   clang-format-7
+                   clang-format-7.0
+                   clang-format-6.0
+                   clang-format-5.0
+                   clang-format-4.0
+                   clang-format-3.9
+                   clang-format-3.8
+                   clang-format-3.7
+                   clang-format-3.6
+                   clang-format-3.5
+                   clang-format-3.4
+                   clang-format-3.3
+                   clang-format  # least priority
+             DOC "clang-format executable")
+mark_as_advanced(CLANG_FORMAT_EXECUTABLE)
+
+if(CLANG_FORMAT_EXECUTABLE)
+  set(CLANG_FORMAT_FOUND TRUE)
+else()
+  set(CLANG_FORMAT_FOUND FALSE)
+  return()
+endif()
+
+# Extract version from command "clang-format -version"
+execute_process(COMMAND ${CLANG_FORMAT_EXECUTABLE} -version
+                OUTPUT_VARIABLE clang_format_version
+                ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+if(NOT clang_format_version MATCHES "^.*clang-format version .*")
+  # clang_format_version samples:
+  # * clang-format version 3.9.1-4ubuntu3~16.04.1 (tags/RELEASE_391/rc2)
+  # * Alpine clang-format version 8.0.0 (tags/RELEASE_800/final) (based on LLVM 8.0.0)
+  unset(clang_format_version)
+  return()
+endif()
+string(REGEX
+        REPLACE ".*clang-format version ([.0-9]+).*"
+                "\\1"
+                CLANG_FORMAT_VERSION
+                "${clang_format_version}")
+# CLANG_FORMAT_VERSION sample: "3.9.1"
```

Comment:
We may stop at this point. `CLANG_FORMAT_VERSION` is what we care about, no reason to extract major, minor, etc.

Reply:
Removed this and modified the rst doc at the top too